### ISSUE TITLE
ML estimation not working for chisq, invgamma, invgauss

### ISSUE THREAD

**Issue by wleoncio** [ocbe-uio/TruncExpFam#90]
- [x] Fix estimation for chisq (6948ba2a)
- [x] Fix estimation for invgamma (2c2036c6)
- [x] Fix estimation for invgauss (06eb8c45)

**Comment by wleoncio** [comment]
Useful reference: https://en.wikipedia.org/wiki/Exponential_family#Table_of_distributions

**Comment by rho62** [comment]
Yes, I used that for the implementations. Should be included in our list of refs

Sendt fra min iPad

Den 25. jul. 2022 kl. 09.35 skrev Waldir Leoncio ***@***.***>:

﻿

Useful reference: https://en.wikipedia.org/wiki/Exponential_family#Table_of_distributions

**Comment by wleoncio** [comment]
Hi @rho62,

After reviewing the code for the invgauss and the ML estimation functions several times, I couldn't find anything to improve. The peculiarity of this distribution is that etas must be both negative or positive, so that operations like `sqrt(eta[2] / eta[1])` can be properly defined in the Real space. The problem that the ML estimation is currently encountering is that sometimes it will just update eta so that one of its elements is positive. By choosing a lower `delta`, the user is able to avoid this (paying the price of a weaker estimation).

My feeling right now is that this could be better handled by improving `mlEstimationTruncDist()` so that it automatically calculates a proper `delta`, unless it's really important for the user to be able to control both `tol` and `delta`, but this would be something for the future.

In my latest commit, the invgamma issue is worked around by adding the following code to `mlEstimationTruncDist()`:

https://github.com/ocbe-uio/TruncExpFam/blob/06eb8c4577ea6f6827f656f49118d87913c67eeb/R/mlEstimationTruncDist.R#L96-L99

In other words, if the updated eta corresponds to invalid parameters, the estimation quits and suggests a smaller delta. In my tests, this only happens in some cases of the truncated inverse gamma distribution.

What do you think? If this is a proper workaround, even if a temporary one, I can close this issue and focus on the last distro giving problems (beta).

**Comment by rho62** [comment]
Hi Waldir

My understanding is that the estimation does not yield wrong results, but rather halts the estimation with a suggestion on a smaller delta? Albeit not ideal it sounds acceptable.
A some point it should be corrected to provide a more robust estimation, but for now it is acceptable.
You write both about the invGamma and the invGaussian … Is there an issue with both?

/René


Fra: Waldir Leoncio ***@***.***>
Svar til: ocbe-uio/TruncExpFam ***@***.***>
Dato: tirsdag 2. august 2022 kl. 13:59
Til: ocbe-uio/TruncExpFam ***@***.***>
Kopi: Rene Holst ***@***.***>, Mention ***@***.***>
Emne: Re: [ocbe-uio/TruncExpFam] ML estimation not working for chisq, invgamma, invgauss (Issue #90)


Hi @rho62<https://github.com/rho62>,

After reviewing the code for the invgauss and the ML estimation functions several times, I couldn't find anything to improve. The peculiarity of this distribution is that etas must be both negative or positive, so that operations like sqrt(eta[2] / eta[1]) can be properly defined in the Real space. The problem that the ML estimation is currently encountering is that sometimes it will just update eta so that one of its elements is positive. By choosing a lower delta, the user is able to avoid this (paying the price of a weaker estimation).

My feeling right now is that this could be better handled by improving mlEstimationTruncDist() so that it automatically calculates a proper delta, unless it's really important for the user to be able to control both tol and delta, but this would be something for the future.

In my latest commit, the invgamma issue is worked around by adding the following code to mlEstimationTruncDist():

https://github.com/ocbe-uio/TruncExpFam/blob/06eb8c4577ea6f6827f656f49118d87913c67eeb/R/mlEstimationTruncDist.R#L96-L99

In other words, if the updated eta corresponds to invalid parameters, the estimation quits and suggests a smaller delta. In my tests, this only happens in some cases of the truncated inverse gamma distribution.

What do you think? If this is a proper workaround, even if a temporary one, I can close this issue and focus on the last distro giving problems (beta).

**Comment by wleoncio** [comment]
> the estimation does not yield wrong results, but rather halts the estimation with a suggestion on a smaller delta?

That is the proposed behavior, yes. For some reason, the default `delta = 0.33` causes the algo to break before it converges. As a matter of fact, it seems to go past the true values, which is when the deltas start increasing out of control, see the reprex below:

``` r
# Package installation and setup ===============================================
remotes::install_github("ocbe-uio/TruncExpFam", force = TRUE, quiet = TRUE)
library("TruncExpFam")
#> Welcome to TruncExpFam 1.0.0.9006.
#> Please read the documentation on ?TruncExpFam to learn more about the package.
set.seed(5228)

# Sampling and estimating ======================================================
n <- rtruncinvgauss(1e3, m = 3, s = 1, a = 0.5)
mlEstimationTruncDist(n, print.iter = TRUE)
#> Estimating parameters for the invgauss distribution
#> it:  1 delta:  0.0001237 parm: 3.799 0.81 
#> it:  2 delta:  0.0001062 parm: 3.754 0.824 
#> it:  3 delta:  9.164e-05 parm: 3.703 0.839 
#> it:  4 delta:   7.97e-05 parm: 3.645 0.852 
#> it:  5 delta:  7.012e-05 parm: 3.579 0.865 
#> it:  6 delta:  6.283e-05 parm: 3.504 0.878 
#> it:  7 delta:  5.806e-05 parm: 3.419 0.89 
#> it:  8 delta:  5.641e-05 parm: 3.323 0.902 
#> it:  9 delta:  5.928e-05 parm: 3.213 0.913 
#> it: 10 delta:  6.978e-05 parm: 3.088 0.925 
#> it: 11 delta:  9.503e-05 parm: 2.944 0.936 
#> it: 12 delta:  0.0001528 parm: 2.777 0.949 
#> it: 13 delta:  0.0002925 parm: 2.583 0.963 
#> it: 14 delta:  0.0006722 parm: 2.354 0.982 
#> it: 15 delta:   0.001892 parm: 2.084 1.008 
#> it: 16 delta:   0.006804 parm: 1.765 1.05 
#> it: 17 delta:    0.03392 parm: 1.393 1.128 
#> it: 18 delta:     0.2782 parm: 0.975 1.305 
#> it: 19 delta:      5.611 parm: 0.542 1.856
#> Error in mlEstimationTruncDist(n, print.iter = TRUE): Failed to converge. Try a smaller delta
mlEstimationTruncDist(n, print.iter = TRUE, delta = 0.1)
#> Estimating parameters for the invgauss distribution
#> it:  1 delta:  1.136e-05 parm: 3.799 0.81 
#> it:  2 delta:  1.085e-05 parm: 3.786 0.814 
#> it:  3 delta:  1.036e-05 parm: 3.771 0.818 
#> it:  4 delta:  9.904e-06 parm: 3.757 0.823
#> [1] 3.7410489 0.8270258
```

<sup>Created on 2022-08-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

This has only been detected for the invgauss.

> You write both about the invGamma and the invGaussian … Is there an issue with both?

No, sorry for mentioning "invgamma" on my previous message, that was a typo. The invgamma issue was fixed on commit https://github.com/ocbe-uio/TruncExpFam/commit/2c2036c6c7f1a9853a54cadc3ef9d9232182978c, which corrected an object in the `getGradETinv()` method.

**Comment by rho62** [comment]
Hi Waldir

You wrote: «The peculiarity of this distribution is that etas must be both negative or positive»
Well in fact mu and lamda are both positive and hence both etas must be negative. I don’t know if that is what causes the error?

/René

Fra: Waldir Leoncio ***@***.***>
Svar til: ocbe-uio/TruncExpFam ***@***.***>
Dato: tirsdag 2. august 2022 kl. 13:59
Til: ocbe-uio/TruncExpFam ***@***.***>
Kopi: Rene Holst ***@***.***>, Mention ***@***.***>
Emne: Re: [ocbe-uio/TruncExpFam] ML estimation not working for chisq, invgamma, invgauss (Issue #90)


Hi @rho62<https://github.com/rho62>,

After reviewing the code for the invgauss and the ML estimation functions several times, I couldn't find anything to improve. The peculiarity of this distribution is that etas must be both negative or positive, so that operations like sqrt(eta[2] / eta[1]) can be properly defined in the Real space. The problem that the ML estimation is currently encountering is that sometimes it will just update eta so that one of its elements is positive. By choosing a lower delta, the user is able to avoid this (paying the price of a weaker estimation).

My feeling right now is that this could be better handled by improving mlEstimationTruncDist() so that it automatically calculates a proper delta, unless it's really important for the user to be able to control both tol and delta, but this would be something for the future.

In my latest commit, the invgamma issue is worked around by adding the following code to mlEstimationTruncDist():

https://github.com/ocbe-uio/TruncExpFam/blob/06eb8c4577ea6f6827f656f49118d87913c67eeb/R/mlEstimationTruncDist.R#L96-L99

In other words, if the updated eta corresponds to invalid parameters, the estimation quits and suggests a smaller delta. In my tests, this only happens in some cases of the truncated inverse gamma distribution.

What do you think? If this is a proper workaround, even if a temporary one, I can close this issue and focus on the last distro giving problems (beta).
